File: src/IECore/MatrixMultiplyOp.cpp
//////////////////////////////////////////////////////////////////////////
//
//  Copyright (c) 2007-2013, Image Engine Design Inc. All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions are
//  met:
//
//     * Redistributions of source code must retain the above copyright
//       notice, this list of conditions and the following disclaimer.
//
//     * Redistributions in binary form must reproduce the above copyright
//       notice, this list of conditions and the following disclaimer in the
//       documentation and/or other materials provided with the distribution.
//
//     * Neither the name of Image Engine Design nor the names of any
//       other contributors to this software may be used to endorse or
//       promote products derived from this software without specific prior
//       written permission.
//
//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
//  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
//
//////////////////////////////////////////////////////////////////////////

#include <cassert>

#include "boost/mpl/eval_if.hpp"
#include "boost/mpl/or.hpp"
#include "boost/type_traits/is_same.hpp"
#include "boost/utility/enable_if.hpp"

#include "IECore/MatrixMultiplyOp.h"
#include "IECore/SimpleTypedData.h"
#include "IECore/TransformationMatrixData.h"
#include "IECore/ObjectParameter.h"
#include "IECore/CompoundParameter.h"
#include "IECore/Object.h"
#include "IECore/NullObject.h"
#include "IECore/DespatchTypedData.h"

using namespace IECore;
using namespace Imath;
using namespace std;
using namespace boost;
using namespace boost::mpl;

IE_CORE_DEFINERUNTIMETYPED( MatrixMultiplyOp );

static TypeId inputTypes[] =
{
	V3fVectorDataTypeId,
	V3dVectorDataTypeId,
	InvalidTypeId
};

static TypeId matrixTypes[] =
{
	M33fDataTypeId,
	M33dDataTypeId,
	M44fDataTypeId,
	M44dDataTypeId,
	TransformationMatrixfDataTypeId,
	TransformationMatrixdDataTypeId,
	InvalidTypeId
};

MatrixMultiplyOp::MatrixMultiplyOp()
	:	ModifyOp(
		"Performs inplace matrix multiplication on 3D vector Data types.",
		new ObjectParameter(
			"result",
			"Converted Data object.",
			new NullObject(),
			DataTypeId
		),
		new ObjectParameter(
			"object",
			"The vector object that will be transformed by the matrix.",
			new NullObject(),
			&inputTypes[0]
		)
	)
{
	m_matrixParameter = new ObjectParameter(
		"matrix",
		"The matrix or transformation used on the multiplication.",
		new M44fData(),
		&matrixTypes[0]
	);

	parameters()->addParameter( m_matrixParameter );
}

MatrixMultiplyOp::~MatrixMultiplyOp()
{

}

ObjectParameter * MatrixMultiplyOp::matrixParameter()
{
	return m_matrixParameter.get();
}

const ObjectParameter * MatrixMultiplyOp::matrixParameter() const
{
	return m_matrixParameter.get();
}

struct MultiplyFunctor
{
	typedef void ReturnType;

	DataPtr data;
	ConstObjectPtr matrix;

	template< typename T, typename U >
	void multiply33( U * data, const T &matrix )
	{
		assert( data );
		
		GeometricData::Interpretation mode = data->getInterpretation();
		typename U::ValueType::iterator beginIt = data->writable().begin();
		typename U::ValueType::iterator endIt = data->writable().end();
		if ( mode==GeometricData::Point || mode==GeometricData::Vector )
		{
			for ( typename U::ValueType::iterator it = beginIt; it != endIt; it++ )
			{
				*it *= matrix;
			}
		}
		else if ( mode == GeometricData::Normal )
		{
			T m = matrix.inverse();
			m.transpose();
			for ( typename U::ValueType::iterator it = beginIt; it != endIt; it++ )
			{
				*it *= matrix;
			}
		}
	}

	template< typename T, typename U >
	void multiply( U * data, const T &matrix )
	{
		assert( data );
		
		GeometricData::Interpretation mode = data->getInterpretation();
		typename U::ValueType::iterator beginIt = data->writable().begin();
		typename U::ValueType::iterator endIt = data->writable().end();
		if ( mode == GeometricData::Point )
		{
			for ( typename U::ValueType::iterator it = beginIt; it != endIt; it++ )
			{
				*it *= matrix;
			}
		}
		else if ( mode == GeometricData::Vector )
		{
			for ( typename U::ValueType::iterator it = beginIt; it != endIt; it++ )
			{
				matrix.multDirMatrix( *it, *it );
			}
		}
		else if ( mode == GeometricData::Normal )
		{
			T m = matrix.inverse();
			m.transpose();
			for ( typename U::ValueType::iterator it = beginIt; it != endIt; it++ )
			{
				m.multDirMatrix( *it, *it );
			}
		}
	}

	template<typename U>
	void operator() ( U * data )
	{
		assert( data );

		switch ( matrix->typeId() )
		{
		case M33fDataTypeId:
			multiply33<M33f, U>( data, boost::static_pointer_cast< const M33fData >( matrix )->readable() );
			break;
		case M33dDataTypeId:
			multiply33<M33d, U>( data, boost::static_pointer_cast< const M33dData >( matrix )->readable() );
			break;
		case M44fDataTypeId:
			multiply<M44f, U>( data, boost::static_pointer_cast< const M44fData >( matrix )->readable() );
			break;
		case M44dDataTypeId:
			multiply<M44d, U>( data, boost::static_pointer_cast< const M44dData >( matrix )->readable() );
			break;
		case TransformationMatrixfDataTypeId:
			multiply<M44f, U>( data, boost::static_pointer_cast< const TransformationMatrixfData >( matrix )->readable().transform() );
			break;
		case TransformationMatrixdDataTypeId:
			multiply<M44d, U>( data, boost::static_pointer_cast< const TransformationMatrixdData >( matrix )->readable().transform() );
			break;
		default:
			throw InvalidArgumentException( "Data supplied is not a known matrix type." );
		}
	}

};

void MatrixMultiplyOp::modify( Object * toModify, const CompoundObject * operands )
{
	Data *data = static_cast< Data * >( toModify );
	MultiplyFunctor func = { data, m_matrixParameter->getValue() };
	despatchTypedData< MultiplyFunctor, TypeTraits::IsFloatVec3VectorTypedData >( data, func );
}

File: test/IECore/MatrixMultiplyOp.py
##########################################################################
#
#  Copyright (c) 2007-2013, Image Engine Design Inc. All rights reserved.
#
#  Redistribution and use in source and binary forms, with or without
#  modification, are permitted provided that the following conditions are
#  met:
#
#     * Redistributions of source code must retain the above copyright
#       notice, this list of conditions and the following disclaimer.
#
#     * Redistributions in binary form must reproduce the above copyright
#       notice, this list of conditions and the following disclaimer in the
#       documentation and/or other materials provided with the distribution.
#
#     * Neither the name of Image Engine Design nor the names of any
#       other contributors to this software may be used to endorse or
#       promote products derived from this software without specific prior
#       written permission.
#
#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
#  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
#  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
#  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
#  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
#  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
#  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
#  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
#  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
#  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
#  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#
##########################################################################

import unittest
import os.path
import imath

import IECore

class TestMultiplyMatrixOp( unittest.TestCase ) :

	def testMultiplication( self ) :
		vectorTypes = [
			IECore.V3fVectorData( [ imath.V3f(1), imath.V3f(2), imath.V3f(3) ], IECore.GeometricData.Interpretation.Vector ),
			IECore.V3dVectorData( [ imath.V3d(1), imath.V3d(2), imath.V3d(3) ], IECore.GeometricData.Interpretation.Vector ),
		]
		matrixTypes = [
			IECore.M33fData( imath.M33f() * 3 ),
			IECore.M33dData( imath.M33d() * 3 ),
			IECore.M44fData( imath.M44f().scale( imath.V3f(3) ) ),
			IECore.M44dData( imath.M44d().scale( imath.V3d(3) ) ),
			IECore.TransformationMatrixfData( IECore.TransformationMatrixf( imath.V3f( 3 ), imath.Eulerf(), imath.V3f( 0 ) ) ),
			IECore.TransformationMatrixdData( IECore.TransformationMatrixd( imath.V3d( 3 ), imath.Eulerd(), imath.V3d( 0 ) ) ),
		]
		for vector in vectorTypes:

			targetVector = vector.copy()
			for i in xrange( len( targetVector) ):
				targetVector[ i ] = targetVector[ i ] * 3

			for matrix in matrixTypes:
				res = IECore.MatrixMultiplyOp()( object = vector.copy(), matrix = matrix )
				if res == targetVector:
					continue
				raise Exception, "Error testing vector " + str(type(vector)) + " against matrix " + str(type(matrix)) + ". Resulted " + str( res )

	def testInterpretations( self ) :

		v = IECore.V3fVectorData( [ imath.V3f( 1 ), imath.V3f( 2 ), imath.V3f( 3 ) ], IECore.GeometricData.Interpretation.Point )
		o = IECore.MatrixMultiplyOp()

		# as points
		vt = o( object = v.copy(), matrix = IECore.M44fData( imath.M44f().translate( imath.V3f( 1, 2, 3 ) ) ) )
		for i in range( v.size() ) :
			self.assertEqual( vt[i], v[i] + imath.V3f( 1, 2, 3 ) )

		# as vectors
		v2 = v.copy()
		v2.setInterpretation( IECore.GeometricData.Interpretation.Vector )
		vt = o( object = v2, matrix = IECore.M44fData( imath.M44f().translate( imath.V3f( 1, 2, 3 ) ) ) )
		for i in range( v.size() ) :
			self.assertEqual( vt[i], v[i] )

		vt = o( object = v2, matrix = IECore.M44fData( imath.M44f().scale( imath.V3f( 1, 2, 3 ) ) ) )
		for i in range( v.size() ) :
			self.assertEqual( vt[i], v[i] * imath.V3f( 1, 2, 3 ) )

		# as normals
		v3 = v.copy()
		v3.setInterpretation( IECore.GeometricData.Interpretation.Normal )
		vt = o( object = v3, matrix = IECore.M44fData( imath.M44f().translate( imath.V3f( 1, 2, 3 ) ) ) )
		for i in range( v.size() ) :
			self.assertEqual( vt[i], v[i] )

		vt = o( object = v3, matrix = IECore.M44fData( imath.M44f().scale( imath.V3f( 1, 2, 3 ) ) ) )
		for i in range( v.size() ) :
			self.assertNotEqual( vt[i], v[i] * imath.V3f( 1, 2, 3 ) )

		# nothing happens for numeric
		v4 = v.copy()
		v4.setInterpretation( IECore.GeometricData.Interpretation.None )
		vt = o( object = v4, matrix = IECore.M44fData( imath.M44f().translate( imath.V3f( 1, 2, 3 ) ) ) )
		for i in range( v.size() ) :
			self.assertEqual( vt[i], v[i] )

		vt = o( object = v4, matrix = IECore.M44fData( imath.M44f().scale( imath.V3f( 1, 2, 3 ) ) ) )
		for i in range( v.size() ) :
			self.assertEqual( vt[i], v[i] )

		# nothing happens for color
		v5 = v.copy()
		v5.setInterpretation( IECore.GeometricData.Interpretation.Color )
		vt = o( object = v5, matrix = IECore.M44fData( imath.M44f().translate( imath.V3f( 1, 2, 3 ) ) ) )
		for i in range( v.size() ) :
			self.assertEqual( vt[i], v[i] )

		vt = o( object = v5, matrix = IECore.M44fData( imath.M44f().scale( imath.V3f( 1, 2, 3 ) ) ) )
		for i in range( v.size() ) :
			self.assertEqual( vt[i], v[i] )

if __name__ == "__main__":
        unittest.main()
